### ISSUE TITLE
Consider all b2_null{Type}Id values as 0

### ISSUE THREAD

**Issue by behcpl** [erincatto/box2c#109]
Taken from include/box2d/id.h various null ids are defined as {-1, -1, 0}. This has different binary representation than nullptr. Some bits are 1 while with nullptr all bits are 0. While there is no big difference in C/C++, because every field/variable must be initialized (unless you want to `memset` 0 the entire structure), other languages might use their own policies. For example in C# `struct` cannot have custom default (no parameters) constructor, the compiler automatically provides a default constructor that initializes all fields to their default values and these are 0 for `int`/`uint`. So with C# binding, the current approach would be much more cumbersome to use and error-prone because you couldn't rely on the default compiler behavior.

**Comment by erincatto** [comment]
I thought about this. How would C# deal with all the def structures, such as `b2BodyDef`? Some values defaulting to zero are problematic, such as gravity scale.

**Comment by behcpl** [comment]
I would do something like this:
```csharp
public struct b2BodyDef
{
    // common pattern in C#, I'm not sure if native b2_defaultBodyDef would be available here in any form
    public static readonly b2BodyDef Default = new b2BodyDef
    {
        //...
        angularVelocity = 0.0f,
        angularDamping = 0.0f,
        gravityScale = 1.0f,
        linearDamping = 0.0f,
        //...
    };
    
    //...
    public float angularVelocity;
    public float linearDamping;
    public float angularDamping;
    public float gravityScale;
    //...
}

public class UseCases
{
    private b2BodyDef localField; //unless explicitly set, it's same as default
    
    public void Method()
    {
        b2BodyDef bd0; //variables are uninitialized, unlike class/struct fields
        float vel = bd0.angularVelocity; //compiler error
        
        // wrong, all 3 zeroed out
        b2BodyDef bd1 = localField;
        b2BodyDef bd2 = default;
        b2BodyDef bd3 = new();
        
        // ok, defaults we wanted
        b2BodyDef bd4 = b2BodyDef.Default;
        
        // ok, preferred way, override on demand 
        b2BodyDef bd5 = b2BodyDef.Default with { linearDamping = 0.5f };
        
        // wrong, gravityScale == 0, must set everything
        b2BodyDef bd6 = new() { linearDamping = 0.5f };
    }
}
```

Of course we can apply same solution for Id structs, and make them have `Null` property, but I think it's a different use case.
I think structs like `b2BodyDef` would be very rarely used as class fields (I would expect mostly parameters or variables) while struct Ids could be in many places in user code.

**Comment by erincatto** [comment]
Okay, that is a relief actually.

I'm considering making the ids be a uint64 with 0 as null. I could still wrap the uint64 in a struct for type safety, but I will likely put the type as some bits in the uint64 as well.

**Comment by behcpl** [comment]
I think using struct for type safety was a good call (whether with single uint64 or with multiple fields). I find no good use case to have some universal Id type (in user code) that can point to body, shape or joint, it would be equivalent to using void*.

With compiler ensured type safety encoding type inside id value would be redundant, but having some bits that defines joint type for b2JointId (and shape type for b2ShapeId) might allow some optimizations so it's worth investigating.

**Comment by erincatto** [comment]
Unlike pointers, ids cannot be forward declared when used in a user header file. This is why I have id.h as a separate light-weight header. There still might be situations where including id.h in a user header is undesirable. So having the void* equivalent might be useful as an unwrapped uint64.

```c
#include "box2d/id.h"
struct UserStruct
{
  b2BodyId bodyId;
};
```

I suppose a user could use a pointer to an id in a header along with a forward declaration. This would likely require a wasteful small allocation of the id itself.

```c
typedef struct b2BodyId b2BodyId;
struct UserStruct
{
  b2BodyId* bodyIdPtr;
};
```

**Comment by erincatto** [comment]
Implemented this in #118. Thanks for your feedback!
